elf_reader.go

```diff
@@ -479,76 +475,118 @@ func mapSpecFromBTF(spec *btf.Spec, name string) (*MapSpec, error) {
 	if err != nil {
 		return nil, fmt.Errorf("can't get size of BTF value: %w", err)
 	}
-	valueSize := uint32(size)
 
-	var (
-		mapType, flags, maxEntries uint32
-		pinType                    PinType
-	)
+	mapSpec := &MapSpec{
+		Name:      SanitizeName(name, -1),
+		ValueSize: uint32(size),
+		KeySize:   keySize,
+	}
+
 	for _, member := range btfMapMembers {
 		switch member.Name {
 		case "type":
-			mapType, err = uintFromBTF(member.Type)
+			mapType, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get type: %w", err)
 			}
+			mapSpec.Type = MapType(mapType)
 
 		case "map_flags":
-			flags, err = uintFromBTF(member.Type)
+			flags, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF map flags: %w", err)
 			}
+			mapSpec.Flags = flags
 
 		case "max_entries":
-			maxEntries, err = uintFromBTF(member.Type)
+			maxEntries, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF map max entries: %w", err)
 			}
+			mapSpec.MaxEntries = maxEntries
 
 		case "key_size":
 			if _, isVoid := keyType.(*btf.Void); !isVoid {
 				return nil, errors.New("both key and key_size given")
 			}
 
-			keySize, err = uintFromBTF(member.Type)
+			keySize, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF key size: %w", err)
 			}
+			mapSpec.KeySize = keySize
 
 		case "value_size":
 			if _, isVoid := valueType.(*btf.Void); !isVoid {
 				return nil, errors.New("both value and value_size given")
 			}
 
-			valueSize, err = uintFromBTF(member.Type)
+			valueSize, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get BTF value size: %w", err)
 			}
+			mapSpec.ValueSize = valueSize
 
 		case "pinning":
+			if isInnermap {
+				return nil, fmt.Errorf("inner def can't be pinned")
```

Comment:
Good idea, but can you move this check to `newMapWithBTF` in map.go instead? Then it applies to "manually" created MapSpec as well.

Reply:
I'll consider it.
However, newMapWithBTF does not have a recursive definition, so I can't think of a reason to have it. (isInnermap is the flag I used to view the map recursively.
Could you please explain it to me?

Reply:
`newMapWithBTF` deals with nested maps here: https://github.com/cilium/ebpf/blob/e8edbe10927def099a96acddf0676ddb38b11ccf/map.go#L211-L223

You can add the check for pinning after the `spec.InnerMap == nil` check.